Show "Publish: Immediately" for new drafts by inferring floating date

kadamwhite opened the pull request:
## Description
This addresses the post date label issue described in #7195 with an alternative to the previously-proposed solution in PR #8395.

Rather than forcing a new property into the API to work around the nuances of how dates are stored in the database, we can use what @earnjam and @TimothyBJacobs proposed in https://core.trac.wordpress.org/ticket/39953 and infer that a post's date is "floating" (_i.e._ that it should be published immediately) by comparing the date modified to the listed publish date. In that Trac thread @TimothyBJacobs explained,

> If the date and the modified date are the same, then we're still working on the post because those are both updated at every write. As mentioned we don't have an actual `created_at` date. This would save us having to add the new field. We couldn't figure out a time when those two dates wouldn't match, and in testing it looked like it would work to fulfill Gutenberg's needs.

Note that this works in Gutenberg because on save, we only send back values which have been edited. To fully address this issue throughout WordPress, changes will be needed in Trac to ensure that we apply the same logic when preparing posts to be saved to the database.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
I created this patch with @earnjam at the WCNYC contributor day, and I have tested it by creating several posts to ensure various combinations of draft and auto-draft correctly reflect manually-specified dates and the Immediately placeholder. To reproduce these tests,

1. Create a new post and look at the date label. It should display "Immediately."
2. Update the title (or any other non-date field) and save a draft. Post should still display Publish: "Immediately."
3. Test publish _vs_ manually-specify date:
  - To test that date is set properly when publishing, publish the post and ensure the date that is displayed on the front-end and when reloading the post in the editor reflects the time at which the post was published.
  - To test that the date is set properly when manually specifying a date, click "Immediately" and set a manual date at which the post should be published. Reload the page and ensure the provided date still displays.

## Screenshots <!-- if applicable -->
![image](https://user-images.githubusercontent.com/442115/45627438-ef0bfd80-ba5f-11e8-8b40-472c701ba2d3.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix for PostScheduleLabel component.
Adds unit tests for PostScheduleLabel component.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

kadamwhite committed:
Show "Publish: Immediately" for new drafts by inferring floating date

youknowriad commented on packages/editor/src/components/post-schedule/label.js:
```diff
@@ -5,15 +5,31 @@ import { __ } from '@wordpress/i18n';
 
 export default withSelect( ( select ) => {
 	return {
 		date: select( 'core/editor' ).getEditedPostAttribute( 'date' ),
```
What if instead of having to do this check in the component, we do it in the selector `getEditedPostAttribute`  instead returning `null` or `undefined` for the `date` attribute if the condition above is met? I'm wondering if a more global change like that is better as it save us having to repeat the same operation if we have to check again in a separate component if needed.

Alternatively, if we don't want to alter the "date" property value "virtually", should we add a new `editedPostHasFloatingDate` selector or something like that?

tofumatt replied:
That makes a lot of sense to me. Us having to deal with this in the component seems odd (and probably what introduced this regression, right?)

youknowriad replied:
I'm actually leaning towards the second option: the separate `editedPostHasFloatingDate` selector because if normalize the "date" property, it should probably be better normalized before it hits the reducer which might be more complex to achieve consistently (several ways to load posts)

kadamwhite replied:
I think I'm in favor of that option as well, because although in a new WP ticket we do intend to support passing `null` to put a post _back_ into floating date status, given the way the API response is constructed a post object will continue to return a date string in all circumstances, and not handling an alternative virtual type in the store feels cleaner.

kadamwhite committed:
Add isEditedPostDateFloating selector

Move logic to determine whether a post should be considered to have a
floating date from the PostScheduleLabel component into a new selector
isEditedPostDateFloating.

youknowriad commented on packages/editor/src/components/post-schedule/label.js:
```diff
@@ -5,15 +5,16 @@ import { __ } from '@wordpress/i18n';
 export default withSelect( ( select ) => {
 	return {
 		date: select( 'core/editor' ).getEditedPostAttribute( 'date' ),
+		floating: select( 'core/editor' ).isEditedPostDateFloating(),
```
minor: maybe isFloating to clarify that it's a boolean?

youknowriad reviewed (commented):
Thanks for the update, Code looks good, I'll just do some testing a bit after (unless someone beats me to it) and approve.

kadamwhite committed:
Rename "floating" component property to "isFloating" for type clarity

kadamwhite commented:
Good point about the property naming, I have made that change :+1:

youknowriad commented on packages/editor/src/store/selectors.js:
```diff
@@ -427,6 +427,29 @@ export function isEditedPostBeingScheduled( state ) {
+ *
+ * @return {boolean} Whether the edited post has a floating date value.
+ */
+export function isEditedPostDateFloating( state ) {
```
I expect this new selector to generate some changes in the docs if you run `npm run docs:build`. Can you commit these changes?

aduth replied:
> I expect this new selector to generate some changes in the docs if you run `npm run docs:build`. Can you commit these changes?

This was never done and there are local changes to commit on master after running the command.

Please review #10234

kadamwhite replied:
:grimacing: my bad, @aduth , hadn't had time to circle back to this and had missed that piece of feedback. I'll know for next time.

aduth replied:
No worries @kadamwhite

youknowriad reviewed (approved):
LGTM 👍

mtias commented:
Thanks for the work here @kadamwhite.
